tests/components/influxdb/test_sensor.py

```diff
@@ -0,0 +1,435 @@
+"""The tests for the InfluxDB sensor."""
+from dataclasses import dataclass
+from typing import Dict, List, Type
+
+from influxdb.exceptions import InfluxDBClientError
+from influxdb_client.rest import ApiException
+import pytest
+from voluptuous import Invalid
+
+from homeassistant.components.influxdb.const import (
+    API_VERSION_2,
+    DEFAULT_API_VERSION,
+    DOMAIN,
+    TEST_QUERY_V1,
+    TEST_QUERY_V2,
+)
+from homeassistant.components.influxdb.sensor import PLATFORM_SCHEMA
+import homeassistant.components.sensor as sensor
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.setup import async_setup_component
+
+from tests.async_mock import MagicMock, patch
+
+INFLUXDB_PATH = "homeassistant.components.influxdb"
+INFLUXDB_CLIENT_PATH = f"{INFLUXDB_PATH}.sensor.InfluxDBClient"
+INFLUXDB_SENSOR_PATH = f"{INFLUXDB_PATH}.sensor"
+
+BASE_V1_CONFIG = {}
+BASE_V2_CONFIG = {
+    "api_version": API_VERSION_2,
+    "organization": "org",
+    "token": "token",
+}
+
+BASE_V1_QUERY = {
+    "queries": [
+        {
+            "name": "test",
+            "measurement": "measurement",
+            "where": "where",
+            "field": "field",
+        }
+    ],
+}
+BASE_V2_QUERY = {"queries_flux": [{"name": "test", "query": "query"}]}
+
+
+@dataclass
+class Record:
+    """Record in a Table."""
+
+    values: Dict
+
+
+@dataclass
+class Table:
+    """Table in an Influx 2 resultset."""
+
+    records: List[Type[Record]]
+
+
+@pytest.fixture(name="mock_client")
+def mock_client_fixture(request):
+    """Patch the InfluxDBClient object with mock for version under test."""
+    if request.param == API_VERSION_2:
+        client_target = f"{INFLUXDB_CLIENT_PATH}V2"
+    else:
+        client_target = INFLUXDB_CLIENT_PATH
+
+    with patch(client_target) as client:
+        yield client
+
+
+@pytest.fixture(autouse=True)
+def mock_influx_recorder_setup():
+    """Mock the influx recorder setup since the sensor is actually independent."""
+    with patch(f"{INFLUXDB_PATH}.setup", return_value=True) as mock_setup:
```

Comment:
We should probably not patch the component setup, but let it set up the integration instead.

Can we just patch the library instead? The tests will be more robust if we don't patch the integration code.

Reply:
Right now the sensor platform doesn't really depend on the main component but that might change in the future if we refactor.

Reply:
Yea I suppose I can do that. I just kind of hate that the sensor depends on the main component setup. I actually had no idea there was a dependence until I started writing these tests. When you configure the sensor you put in all the details to connect to an InfluxDB instance and an independent connection to the Influx client is created just for the sensor. 

So although the sensor is part of the InfluxDB Integration the two components are completely independent. The sensor can query data from whatever InfluxDB client it wants regardless of what one you are writing data to. It seems very odd to me that you *must* be writing data to one in order to query data from another.

But I suppose you're right, for now that dependency does exist so I probably shouldn't be mocking it. Patching the client library in both classes should accomplish the same thing. Will just need to adjust base config in the `_setup` method.

Reply:
Ok so what I settled on here is mocking client library for the main platform during the sensor tests as well as mocking `Queue.get` to force it to return `None`. This causes the listener to immediately shutdown after initialization, minimizing its impact on the sensor tests. Let me know what you think

Reply:
> Right now the sensor platform doesn't really depend on the main component but that might change in the future if we refactor.

That's true. Actually my thinking for making this component UI configurable was along those lines. I was thinking you would start by configuring the connection to influx then in the options flow decide whether you want any sensors created from queries and if you want to write out state to it. And then if you decide to write state, what to exclude, include, etc.

Unfortunately I'm discovering in testing that the config entry UI isn't really ready for that yet. Seems like there's no support displaying lists right now in config entry UI which makes it pretty difficult to make the component UI configurable right now, there's a lot of lists. Figured I'd look into that first since that seems like something a lot of people could use.